include/glow/Optimizer/GraphOptimizer/Pipeline/Pipeline.h

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_OPTIMIZER_GRAPHOPTIMIZER_GRAPHOPTIMIZER_PIPELINES_PIPELINES_H
+#define GLOW_OPTIMIZER_GRAPHOPTIMIZER_GRAPHOPTIMIZER_PIPELINES_PIPELINES_H
+
+#include "glow/Optimizer/GraphOptimizer/CompilationContext.h"
+#include "glow/Support/Support.h"
+
+#include <bitset>
+
+namespace glow {
+
+/// Define an enum to identify all FunctionPasses, used to declare inside a
+/// FunctionPassConfig for a Pipeline.
+enum class FunctionPassID {
+#define FUN_PASS(PASS_NAME) PASS_NAME,
+#include "glow/Optimizer/GraphOptimizer/FunctionPasses.def"
+};
+
+/// Specifies convergence mode for a FunctionPass.
+enum class ConvergenceMode {
+  OnePass,         // Run a single pass over the Function.
+  UntilFixedPoint, // Run the pass over the Function until a fixed point is
+                   // reached.
+};
+
+/// Specifies a configuration for running a FunctionPass when used in a
+/// FunctionPassPipeline.
+class FunctionPassConfig {
+public:
+  /// Specifies whether the pass requires DCE.
+  enum class DCERequiredMode {
```

Comment:
Nit: Overall, I'm still not a very big fan of making DCE so special. But it is probably OK. 

Maybe to make it less special we could have here something like `PassID prePass`, which is the pass to always run before the current one? It can be initialized to DCE pass ID by default, but could be something else if needed. `getDCERequiredMode` would become `getPrePass` then. And some checks will become less DCE centric. Don't know if it is an improvement though, as it the only actual use-case for now will be using DCE and my proposal could kind of make it less obvious ;-) WDYT?

Reply:
I thought about this, but I was a bit weary of it because I didn't want the pipeline to have the potential of accidentally infinite looping. E.g. you could imagine a circular dependence on passes via `prePass` which isn't trivial to detect. For now I made it DCE because you can always check that DCE does not require DCE before.

Reply:
> E.g. you could imagine a circular dependence on passes via prePass which isn't trivial to detect.

Pass demanding itself as a prePass is very easy to detect (pretty much like you do for DCE now), but any non-obvious recursion would require more effort to detect it of course (e.g. one could check the list of passes executed so far and check for repeating never ending patterns ;-). 

As I said, I don't think it is a big issue. I jut don't like the fact that DCE is so special-cased at the moment.

Reply:
Yeah the non-obvious recursion is what I was thinking of. There aren't other passes which depend on others right now anyway other than DCE, so even if we added such an ability it's not currently useful, and like you said it makes the use-case less obvious. I'd prefer to just stick with DCE special cased for now and if we find its useful to expand it in the future we can.